src/CPUActivityAgent.cpp

```diff
@@ -299,34 +241,55 @@ namespace geopm
         m_do_send_policy = false;
         m_do_write_batch = false;
 
-        if (m_qm_max_rate.empty()) {
-            for (auto it = in_policy.begin() + M_POLICY_FIRST_UNCORE_FREQ;
-                 it != in_policy.end() && std::next(it) != in_policy.end();
-                 std::advance(it, 2)) {
-
-                auto uncore_freq = (*it);
-                auto max_mem_bw = *(it + 1);
-                if (!std::isnan(uncore_freq)) {
-                    // Not valid to have NAN max mem bw for uncore freq.
-                    GEOPM_DEBUG_ASSERT(!std::isnan(max_mem_bw),
-                                       "mapped CPU_UNCORE_FREQUENCY with no max memory bandwidth assigned.");
-                    m_qm_max_rate[uncore_freq] = max_mem_bw;
-                }
-            }
+        // Calculate new frequency range values
+        double f_core_range = m_freq_core_max - m_freq_core_efficient;
+        double f_uncore_range = m_freq_uncore_max - m_freq_uncore_efficient;
 
-            if (m_qm_max_rate.empty()) {
-                throw Exception("CPUActivityAgent::" + std::string(__func__) +
-                                "(): CPUActivityAgent policy did not contain" +
-                                " memory bandwidth characterization.",
-                                GEOPM_ERROR_INVALID, __FILE__, __LINE__);
-            }
+        double phi = in_policy[M_POLICY_CPU_PHI];
+
+        // Default phi = 0.5 case is full fe to fmax range
+        // Core
+        m_resolved_f_core_max = m_freq_core_max;
+        m_resolved_f_core_efficient = m_freq_core_efficient;
+        // Uncore
+        m_resolved_f_uncore_max = m_freq_uncore_max;
+        m_resolved_f_uncore_efficient = m_freq_uncore_efficient;
+
+        // If phi is not 0.5 we move into the energy or performance biased behavior
+        if (phi > 0.5) {
```

Comment:
Q: can phi change during the execution? If not, then I'm thinking this code should be executed once (i.e. moved to the init phase).

Reply:
In our current usage Phi does not change during execution, however we want the capability to change the phi value passed to child agents in the future for potential multi-node algorithms.

Phi is provided via the policy, so I'm not sure how we'd move it to init since we have to call validate_policy.  Previously this logic on checking phi was handled in validate_policy.

Reply:
Right, good point (init wouldn't be the right place for it). So I would recommend leaving it in validate policy until there is a need to accommodate potential multi-node algorithms in the future.

Reply:
Validate policy is const (can't modify member variables).  Previously we worked around this by modifying the in_policy values for max freq & efficient freq, but those have been pulled out of the policy in this PR (using the GPU IOGroup & ConstConfig).

Reply:
Well, then :)